File: Leetcode/JavaScript Solutions/JewelsandStones.js
/*
LeetCode Problem 771. Jewels and Stones
Link: https://leetcode.com/problems/jewels-and-stones/
Written by: Mostofa Adib Shakib
Language: JavaScript
*/


# Solution 1
# Time Complexity: O(n)
# Space Complexity: O(n)

var numJewelsInStones = function(J, S) {
    var hashmap = {};
    var count = 0;
    
    for (var i in J){
        if(!hashmap.hasOwnProperty(i)){
            hashmap[J[i]] = 1
        }
    }
    
    for (var i in S){
        if(hashmap.hasOwnProperty(S[i])){
            count++;
        }
    }
    return count;
};


# Solution 2
# Time Complexity: O(n^2)
# Space Complexity: O(n)

var numJewelsInStones = function(J, S) {
    const array = J.split("")
    var count = 0;
    
    for (var i in S){
        if(array.includes(S[i])){
            count++;
        }
    }
    return count;
};

File: Leetcode/Python Solutions/Arrays/JewelsandStones.py
"""
LeetCode Problem: 771. Jewels and Stones
Link: https://leetcode.com/problems/jewels-and-stones/
Language: Python
Written by: Mostofa Adib Shakib
"""

class Solution:
    def numJewelsInStones(self, J: str, S: str) -> int:
        hashmap = {}
        for i in J:
            if i not in hashmap:
                hashmap[i] = 1
        count = 0
        
        for i in S:
            if i in hashmap:
                count += 1
                
        return count